File: webmcp/test/globs.test.js
const { routeGlobToRegex, pathMatchesRoute } = require("../src/globs");

describe("route globs (spec: Route matching)", () => {
  test("* matches exactly one segment", () => {
    expect(pathMatchesRoute("/users/42", "/users/*")).toBe(true);
    expect(pathMatchesRoute("/users/42/edit", "/users/*")).toBe(false);
    expect(pathMatchesRoute("/users", "/users/*")).toBe(false);
  });

  test("** as its own segment matches zero or more segments", () => {
    expect(pathMatchesRoute("/admin", "/admin/**")).toBe(true);
    expect(pathMatchesRoute("/admin/users", "/admin/**")).toBe(true);
    expect(pathMatchesRoute("/admin/users/42/edit", "/admin/**")).toBe(true);
    expect(pathMatchesRoute("/a/b", "/a/**/b")).toBe(true);
    expect(pathMatchesRoute("/a/x/b", "/a/**/b")).toBe(true);
  });

  test("** glued into a segment degrades to in-segment *", () => {
    expect(pathMatchesRoute("/foobar", "/foo**")).toBe(true);
    expect(pathMatchesRoute("/foo/bar", "/foo**")).toBe(false);
  });

  test(":param normalizes to one segment", () => {
    expect(
      pathMatchesRoute("/us/en/cat/desk-chairs", "/us/en/cat/:categorySlug"),
    ).toBe(true);
    expect(pathMatchesRoute("/us/en/cat/a/b", "/us/en/cat/:categorySlug")).toBe(
      false,
    );
  });

  test("trailing slashes and query/fragment are normalized away", () => {
    expect(pathMatchesRoute("/search/", "/search")).toBe(true);
    expect(pathMatchesRoute("/search?q=x", "/search")).toBe(true);
  });

  test("root route", () => {
    expect(pathMatchesRoute("/", "/")).toBe(true);
    expect(pathMatchesRoute("/x", "/")).toBe(false);
  });

  test("catch-all", () => {
    expect(pathMatchesRoute("/anything/at/all", "/**")).toBe(true);
    expect(pathMatchesRoute("/", "/**")).toBe(true);
  });

  test("regex source is embeddable", () => {
    const re = new RegExp(routeGlobToRegex("/us/en/p/:productSlug").source);
    expect(re.test("/us/en/p/goersnygg-40504193")).toBe(true);
  });
});

File: go/webmcp/globs_test.go
package webmcp

import (
	"regexp"
	"testing"
)

func TestPathMatchesRouteOneSegmentStar(t *testing.T) {
	if !pathMatchesRoute("/users/42", "/users/*") {
		t.Fatal("expected match")
	}
	if pathMatchesRoute("/users/42/edit", "/users/*") {
		t.Fatal("star must not cross a slash")
	}
	if pathMatchesRoute("/users", "/users/*") {
		t.Fatal("star needs a segment")
	}
}

func TestPathMatchesRouteGlobstar(t *testing.T) {
	cases := []struct {
		path, route string
		want        bool
	}{
		{"/admin", "/admin/**", true},
		{"/admin/users", "/admin/**", true},
		{"/admin/users/42/edit", "/admin/**", true},
		{"/a/b", "/a/**/b", true},
		{"/a/x/b", "/a/**/b", true},
	}
	for _, tc := range cases {
		if got := pathMatchesRoute(tc.path, tc.route); got != tc.want {
			t.Errorf("pathMatchesRoute(%q, %q) = %v, want %v", tc.path, tc.route, got, tc.want)
		}
	}
}

func TestPathMatchesRouteGluedGlobstarDegradesToInSegmentStar(t *testing.T) {
	if !pathMatchesRoute("/foobar", "/foo**") {
		t.Fatal("glued ** should match inside the segment")
	}
	if pathMatchesRoute("/foo/bar", "/foo**") {
		t.Fatal("glued ** must not cross a slash")
	}
}

func TestPathMatchesRouteParamIsOneSegment(t *testing.T) {
	if !pathMatchesRoute("/us/en/cat/desk-chairs", "/us/en/cat/:categorySlug") {
		t.Fatal("expected match")
	}
	if pathMatchesRoute("/us/en/cat/a/b", "/us/en/cat/:categorySlug") {
		t.Fatal(":param is one segment")
	}
}

func TestPathMatchesRouteNormalizesSlashQueryFragment(t *testing.T) {
	if !pathMatchesRoute("/search/", "/search") {
		t.Fatal("trailing slash")
	}
	if !pathMatchesRoute("/search?q=x", "/search") {
		t.Fatal("query")
	}
}

func TestPathMatchesRouteRoot(t *testing.T) {
	if !pathMatchesRoute("/", "/") {
		t.Fatal("root")
	}
	if pathMatchesRoute("/x", "/") {
		t.Fatal("root is exact")
	}
}

func TestPathMatchesRouteCatchAll(t *testing.T) {
	if !pathMatchesRoute("/anything/at/all", "/**") {
		t.Fatal("catch-all path")
	}
	if !pathMatchesRoute("/", "/**") {
		t.Fatal("catch-all root")
	}
}

func TestRouteGlobRegexSourceIsEmbeddable(t *testing.T) {
	re, err := regexp.Compile(routeGlobRegexSource("/us/en/p/:productSlug"))
	if err != nil {
		t.Fatal(err)
	}
	if !re.MatchString("/us/en/p/goersnygg-40504193") {
		t.Fatalf("source %q did not match", routeGlobRegexSource("/us/en/p/:productSlug"))
	}
}
